keras/layers/variational.py

```diff
@@ -0,0 +1,105 @@
+    regularizer_scale: By default the regularization is already properly
+        scaled if you use binary or categorical crossentropy cost functions.
+        In most cases this regularizers should be kept fixed at one.
+
+    Notes:
+    ------
+    During training we output ONE sample from the reparameterized
+    distribution given by "mean + std * noise".
+    For testing, the output of this layer is deterministic given by "mean".
+
+    References:
+    -----------
+    - [Auto-Encoding Variational Bayes](http://arxiv.org/pdf/1312.6114v10.pdf)
+
+    """
+    def __init__(self, output_dim, batch_size, init='glorot_uniform',
+                 activation='tanh',
+                 weights=None, input_dim=None, regularizer_scale=1,
+                 prior_mean=0, prior_logsigma=1, **kwargs):
+        self.prior_mean = prior_mean
+        self.prior_logsigma = prior_logsigma
+        self.regularizer_scale = regularizer_scale
+        self.batch_size = batch_size
+        self.init = initializations.get(init)
+        self.activation = activations.get(activation)
+        self.output_dim = output_dim
+        self.initial_weights = weights
+        self.input_dim = input_dim
+        if self.input_dim:
+            kwargs['input_shape'] = (self.input_dim,)
+        self.input = K.placeholder(ndim=2)
+        super(VariationalDense, self).__init__(**kwargs)
+
+    def build(self):
+        input_dim = self.input_shape[-1]
+
+        self.W_mean = self.init((input_dim, self.output_dim))
+        self.b_mean = K.zeros((self.output_dim,))
+        self.W_logsigma = self.init((input_dim, self.output_dim))
+        self.b_logsigma = K.zeros((self.output_dim,))
+
+        self.trainable_weights = [self.W_mean, self.b_mean, self.W_logsigma,
+                                  self.b_logsigma]
+
+        self.regularizers = []
+        reg = self.get_variational_regularization(self.get_input())
+        self.regularizers.append(reg)
+
+    def get_variational_regularization(self, X):
+        mean = self.activation(K.dot(X, self.W_mean) + self.b_mean)
+        logsigma = self.activation(K.dot(X, self.W_logsigma) + self.b_logsigma)
+        return GaussianKL(mean, logsigma,
+                          regularizer_scale=self.regularizer_scale,
+                          prior_mean=self.prior_mean,
+                          prior_logsigma=self.prior_logsigma)
+
+    def get_mean_logsigma(self, X):
+        mean = self.activation(K.dot(X, self.W_mean) + self.b_mean)
+        logsigma = self.activation(K.dot(X, self.W_logsigma) + self.b_logsigma)
+        return mean, logsigma
+
+    def _get_output(self, X, train=False):
+        mean, logsigma = self.get_mean_logsigma(X)
+        if train:
+            # infer batch size for theano backend
+            if K._BACKEND == 'theano':
+                eps = K.random_normal((X.shape[0], self.output_dim))
+            else:
+                eps = K.random_normal((self.batch_size, self.output_dim))
```

Comment:
The `shape` argument to `K.random_normal` can be an int32 TensorFlow or Theano tensor. With Theano the easy syntax `K.random_normal((K.shape(X)[0], output_dim))` would work, but it wouldn't work in TF. So you need to construct the int32 shape tensor.

Something like:
```python
shape = K.concatenate([K.shape(X)[0], np.array(output_dim).astype('int32')])
K.random_normal(shape)
```
May need to be adapted.